File: app/src/main/java/me/smbduknow/transport/DrawableUtil.java
package me.smbduknow.transport;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.LightingColorFilter;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

public class DrawableUtil {

    public static BitmapDrawable writeOnDrawable(Context ctx, int drawableId, String text, float angle){
        Bitmap bm = BitmapFactory.decodeResource(ctx.getResources(), drawableId).copy(Bitmap.Config.ARGB_8888, true);

        bm = rotateBitmap(bm, angle);

        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.WHITE);
        paint.setTextSize(18);
        paint.setTextAlign(Paint.Align.CENTER);

        Canvas canvas = new Canvas(bm);
        canvas.drawText(text, bm.getWidth()/2, bm.getHeight()/2+8, paint);

        return new BitmapDrawable(bm);
    }

    public static Bitmap rotateBitmap(Bitmap source, float angle) {
        Matrix matrix = new Matrix();
        matrix.postRotate(angle);
        return Bitmap.createBitmap(source, 0, 0, source.getWidth(), source.getHeight(), matrix, true);
    }

}

File: app/src/main/java/me/smbduknow/transport/commons/DrawableUtil.kt
package me.smbduknow.transport.commons

import android.content.Context
import android.content.res.Resources
import android.graphics.*
import android.os.Build
import android.support.annotation.DrawableRes
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v7.content.res.AppCompatResources
import android.text.TextPaint
import me.smbduknow.transport.R


object DrawableUtil {

    private val drawableCache = mutableMapOf<Int, Bitmap>()

    fun createVehiclePin(ctx: Context, @DrawableRes iconRes: Int, text: String, angle: Float): Bitmap {

        val iconBitmap = getBitmapFromVectorDrawable(ctx, iconRes)

        val bitmap = Bitmap.createBitmap(iconBitmap.width, iconBitmap.height, Bitmap.Config.ARGB_8888)

        val canvas = Canvas(bitmap)

        // draw icon
        val bitmapPaint = Paint().apply {
            isAntiAlias = true
            isFilterBitmap = true
        }
        val matrix = Matrix().apply{
            setRotate(180 + angle, iconBitmap.width / 2f, iconBitmap.height / 2f)
        }
        canvas.drawBitmap(iconBitmap, matrix, bitmapPaint)

        // draw text
        val fontSize = ctx.resources.getDimensionPixelSize(R.dimen.text_pin).toFloat()
        val textPaint = (TextPaint(TextPaint.ANTI_ALIAS_FLAG)).apply {
            style = Paint.Style.FILL
            color = Color.WHITE
            textSize = fontSize
            textAlign = Paint.Align.CENTER
        }

        canvas.drawText(text, (bitmap.width / 2).toFloat(), (bitmap.height / 2 + fontSize/3), textPaint)

        return bitmap
    }

    private fun getBitmapFromVectorDrawable(context: Context, @DrawableRes drawableId: Int): Bitmap {
        val cached = if(drawableCache.containsKey(drawableId)) drawableCache[drawableId] else null
        if(cached != null) return cached

        var drawable = AppCompatResources.getDrawable(context, drawableId)
                ?: throw Resources.NotFoundException("Vehicle icon not found")

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            drawable = DrawableCompat.wrap(drawable).mutate()
        }

        val size = context.resources.getDimensionPixelSize(R.dimen.pin_size)

        val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)

        drawableCache[drawableId] = bitmap

        return bitmap
    }

}
